File: python_benchmark/benchmark.py
# =========================
#   P-RAF: Python Benchmark Worker (Queue/Concurrent Version)
# =========================

import socket
import json
import random
import time
import threading
import queue

ORCHESTRATOR_HOST = '127.0.0.1'
ORCHESTRATOR_PORT = 9000
LANGUAGE_NAME = "Python"
ALGORITHMS = [
    "array_sort", "fibonacci", "game_of_life",
    "matrix_multiplication", "prime_factors", "is_leap_year"
]

result_queue = queue.Queue(maxsize=1000)

def array_sort():
    a = [random.randint(0, 99) for _ in range(40)]
    a.sort()

def fibonacci():
    a, b = 0, 1
    for _ in range(random.randint(20, 35)):
        a, b = b, a + b

def game_of_life():
    w, h, g = 35, 35, 5
    grid = [[random.randint(0, 1) for _ in range(h)] for _ in range(w)]
    for _ in range(g):
        next_grid = [[0] * h for _ in range(w)]
        for x in range(w):
            for y in range(h):
                ln = sum(
                    grid[(x + i) % w][(y + j) % h]
                    for i in (-1, 0, 1) for j in (-1, 0, 1)
                    if not (i == 0 and j == 0)
                )
                il = grid[x][y] == 1
                next_grid[x][y] = 1 if (il and (ln == 2 or ln == 3)) or (not il and ln == 3) else 0
        grid = next_grid

def matrix_multiplication():
    d = 50
    a = [[random.randint(0, 9) for _ in range(d)] for _ in range(d)]
    b = [[random.randint(0, 9) for _ in range(d)] for _ in range(d)]
    res = [[0] * d for _ in range(d)]
    for i in range(d):
        for j in range(d):
            for k in range(d):
                res[i][j] += a[i][k] * b[k][j]

def prime_factors():
    n = random.randint(100000, 500000)
    d = 2
    fac = []
    while d * d <= n:
        if n % d == 0:
            fac.append(d)
            n //= d
        else:
            d += 1
    if n > 1:
        fac.append(n)

def is_leap_year():
    y = random.randint(1, 9999)
    _ = (y % 4 == 0 and (y % 100 != 0 or y % 400 == 0))

# Each algo thread: run forever, push result to queue
def algo_worker(name, algo_func):
    while True:
        start = time.time()
        algo_func()
        duration = time.time() - start
        result_queue.put({
            "lang": LANGUAGE_NAME,
            "algo": name,
            "duration": duration
        })

# Sender thread: keep one socket, send all results, reconnect if needed
def sender():
    while True:
        try:
            with socket.create_connection((ORCHESTRATOR_HOST, ORCHESTRATOR_PORT)) as sock:
                sock_file = sock.makefile("w", encoding="utf-8", newline="\n")
                print(f"✅ Python sender connected to orchestrator.")
                while True:
                    result = result_queue.get()
                    try:
                        sock_file.write(json.dumps(result) + "\n")
                        sock_file.flush()
                    except Exception as e:
                        print(f"⚠️  Python sender write error: {e}. Reconnecting...")
                        break  # Will reconnect outer loop
        except Exception as e:
            print(f"⚠️  Python sender connect error: {e}. Retrying in 2s...")
            time.sleep(2)

if __name__ == "__main__":
    print("✅ Starting P-RAF Python worker (queue version)...")
    algo_funcs = [
        array_sort, fibonacci, game_of_life,
        matrix_multiplication, prime_factors, is_leap_year
    ]
    # Algo threads
    threads = []
    for name, func in zip(ALGORITHMS, algo_funcs):
        t = threading.Thread(target=algo_worker, args=(name, func), daemon=True)
        t.start()
        threads.append(t)
    # Sender thread
    s = threading.Thread(target=sender, daemon=True)
    s.start()
    for t in threads:
        t.join()

File: ruby_benchmark/benchmark.rb
# =========================
#   P-RAF: Ruby Benchmark Worker (Queue/Concurrent Version)
# =========================

require 'socket'
require 'json'
require 'thread'

ORCHESTRATOR_HOST = '127.0.0.1'
ORCHESTRATOR_PORT = 9000
LANGUAGE_NAME = 'Ruby'
ALGORITHMS = %w[array_sort fibonacci game_of_life matrix_multiplication prime_factors is_leap_year]

def array_sort
  a = Array.new(40) { rand(100) }
  a.sort!
end

def fibonacci
  a, b = 0, 1
  (rand(20..35)).times { a, b = b, a + b }
end

def game_of_life
  w, h, g = 35, 35, 5
  grid = Array.new(w) { Array.new(h) { rand(2) } }
  g.times do
    new_grid = Array.new(w) { Array.new(h, 0) }
    w.times do |x|
      h.times do |y|
        ln = 0
        (-1..1).each do |i|
          (-1..1).each do |j|
            next if i.zero? && j.zero?
            xi, yj = (x + i) % w, (y + j) % h
            ln += grid[xi][yj]
          end
        end
        il = grid[x][y] == 1
        new_grid[x][y] = (il && (ln == 2 || ln == 3)) || (!il && ln == 3) ? 1 : 0
      end
    end
    grid = new_grid
  end
end

def matrix_multiplication
  d = 50
  a = Array.new(d) { Array.new(d) { rand(10) } }
  b = Array.new(d) { Array.new(d) { rand(10) } }
  res = Array.new(d) { Array.new(d, 0) }
  d.times do |i|
    d.times do |j|
      d.times { |k| res[i][j] += a[i][k] * b[k][j] }
    end
  end
end

def prime_factors
  n = rand(100_000..500_000)
  f = []
  d = 2
  while d * d <= n
    if n % d == 0
      f << d
      n /= d
    else
      d += 1
    end
  end
  f << n if n > 1
end

def is_leap_year
  y = rand(1..9999)
  leap = (y % 4 == 0 && (y % 100 != 0 || y % 400 == 0))
end

ALGO_FUNCTIONS = {
  "array_sort" => method(:array_sort),
  "fibonacci" => method(:fibonacci),
  "game_of_life" => method(:game_of_life),
  "matrix_multiplication" => method(:matrix_multiplication),
  "prime_factors" => method(:prime_factors),
  "is_leap_year" => method(:is_leap_year)
}

# Shared thread-safe queue for results
RESULT_QUEUE = Queue.new

def algo_worker(name, algo_proc)
  loop do
    start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    algo_proc.call
    duration = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start
    RESULT_QUEUE << { lang: LANGUAGE_NAME, algo: name, duration: duration }
  end
end

def sender
  loop do
    begin
      puts "🔄 Ruby sender connecting to orchestrator..."
      TCPSocket.open(ORCHESTRATOR_HOST, ORCHESTRATOR_PORT) do |sock|
        puts "✅ Ruby sender connected."
        while true
          payload = RESULT_QUEUE.pop
          begin
            sock.puts(payload.to_json)
          rescue => e
            puts "⚠️  Ruby sender write error: #{e}. Reconnecting..."
            break # Outer loop will reconnect
          end
        end
      end
    rescue => e
      puts "⚠️  Ruby sender connect error: #{e}. Retrying in 2s..."
      sleep 2
    end
  end
end

puts "✅ Starting P-RAF Ruby worker (queue version)..."

threads = []
ALGORITHMS.each do |name|
  threads << Thread.new { algo_worker(name, ALGO_FUNCTIONS[name]) }
end

# Start the sender thread
sender_thread = Thread.new { sender }

threads.each(&:join)
sender_thread.join
